### PR TITLE
Add exhaustive list of error codes and description of metric flag

### PR DESCRIPTION
**Pull request by jake-engelberg**
Thank you for helping make the Dapr documentation better!

**Please follow this checklist before submitting:**
- [x] Commits are signed with Developer Certificate of Origin (DCO - [learn more](https://docs.dapr.io/contributing/contributing-overview/#developer-certificate-of-origin-signing-your-work))
- [ ] [Read the contribution guide](https://docs.dapr.io/contributing/docs-contrib/contributing-docs/)
- [ ] Commands include options for Linux, MacOS, and Windows within codetabs
- [ ] New file and folder names are globally unique
- [ ] Page references use shortcodes instead of markdown or URL links
- [x] Images use HTML style and have alternative text
- [x] Places where multiple code/command options are given have codetabs

In addition, please fill out the following to help reviewers understand this pull request:

## Description

Added the exhaustive set of error codes dapr emits as well as how to enable metric recording for these error codes based on the associated PR below which implements this.

## Issue reference

https://github.com/dapr/dapr/pull/8256

### CODE REVIEW

File: daprdocs/content/en/operations/observability/metrics/metrics-overview.md [hunk 1]
```diff
@@ -70,6 +70,22 @@ spec:
     enabled: false
 ```
 
+## Configuring metrics for error codes
+
+You can enable additional metrics for [Dapr API error codes](https://github.com/dapr/dapr/blob/master/docs/reference/api/error_codes/) by setting `spec.metrics.recordErrorCodes` to `true`. See the specific metrics described in the [Dapr development docs](https://github.com/dapr/dapr/blob/master/docs/development/dapr-metrics.md).
```
Comment (nelson-parente) [inline daprdocs/content/en/operations/observability/metrics/metrics-overview.md]: I think this section should answer:
- why do I want to enable this?
- what happens when I enable this?

Also, mention metrics are collected and maybe add some metric examples

**Review by nelson-parente — APPROVED** [review]

File: daprdocs/content/en/operations/observability/metrics/metrics-overview.md (hunk 1 above)
Comment (hhunter-ms) [reply daprdocs/content/en/operations/observability/metrics/metrics-overview.md]: @jake-engelberg friendly bump

File: daprdocs/content/en/operations/observability/metrics/metrics-overview.md [hunk 2]
```diff
@@ -70,6 +70,38 @@ spec:
     enabled: false
 ```
 
+## Configuring metrics for error codes
+
+You can enable additional metrics for [Dapr API error codes](https://github.com/dapr/dapr/blob/master/docs/reference/api/error_codes/) by setting `spec.metrics.recordErrorCodes` to `true`. Dapr APIs which communicate back to its caller may return standardized error codes. As described in the [Dapr development docs](https://github.com/dapr/dapr/blob/master/docs/development/dapr-metrics.md), a new metric called `error_code_total` will be recorded, which will allow monitoring of error codes triggered by application, code, and category. See [package errorcodes](https://github.com/dapr/dapr/blob/master/pkg/messages/errorcodes/errorcodes.go) for specific codes and categories.
```
Comment (hhunter-ms) [inline daprdocs/content/en/operations/observability/metrics/metrics-overview.md]:
```suggestion
You can enable additional metrics for [Dapr API error codes](https://github.com/dapr/dapr/blob/master/docs/reference/api/error_codes/) by setting `spec.metrics.recordErrorCodes` to `true`. Dapr APIs which communicate back to their caller may return standardized error codes. As described in the [Dapr development docs](https://github.com/dapr/dapr/blob/master/docs/development/dapr-metrics.md), a new metric called `error_code_total` is recorded, which allows monitoring of error codes triggered by application, code, and category. See [the `errorcodes` package](https://github.com/dapr/dapr/blob/master/pkg/messages/errorcodes/errorcodes.go) for specific codes and categories.
```

**Review by hhunter-ms — CHANGES_REQUESTED** [review]
quick review - thank you!

**Review by hhunter-ms — APPROVED** [review]

File: daprdocs/content/en/reference/api/error_codes.md [hunk 3]
```diff
@@ -7,43 +7,150 @@ weight: 1400
...
-| ERR_MALFORMED_REQUEST             | Error with a malformed request.
-| ERR_DIRECT_INVOKE                 | Error in direct invocation.
-| ERR_DESERIALIZE_HTTP_BODY         | Error deserializing an HTTP request body.
-| ERR_SECRET_STORES_NOT_CONFIGURED  | Error that no secret store is configured.
-| ERR_SECRET_STORE_NOT_FOUND        | Error that specified secret store is not found.
-| ERR_HEALTH_NOT_READY              | Error that Dapr is not ready.
-| ERR_METADATA_GET                  | Error parsing the Metadata information.
+The following tables list the error codes returned by Dapr runtime:
+
+### Actors (Building Block)
```
Comment (msfussell) [inline daprdocs/content/en/reference/api/error_codes.md]:
```suggestion
### Actors API
```

File: daprdocs/content/en/reference/api/error_codes.md [hunk 4]
```diff
@@ -7,43 +7,150 @@ weight: 1400
...
+| ERR_ACTOR_TIMER_CREATE           | Error creating a timer for an actor.       |
+| ERR_ACTOR_TIMER_DELETE           | Error deleting a timer for an actor.       |
+| ERR_ACTOR_REMINDER_GET           | Error getting a reminder for an actor.     |
+| ERR_ACTOR_INVOKE_METHOD          | Error invoking a method on an actor.       |
+| ERR_ACTOR_STATE_DELETE           | Error deleting the state for an actor.     |
+| ERR_ACTOR_STATE_GET              | Error getting the state for an actor.      |
+| ERR_ACTOR_STATE_TRANSACTION_SAVE | Error storing actor state transactionally. |
+| ERR_ACTOR_REMINDER_NON_HOSTED    | Error setting reminder for an actor.       |
+
+### Workflows (Building Block)
```
Comment (msfussell) [inline daprdocs/content/en/reference/api/error_codes.md]:
```suggestion
### Workflows API
```

File: daprdocs/content/en/reference/api/error_codes.md [hunk 5]
```diff
@@ -7,43 +7,150 @@ weight: 1400
...
+| ERR_WORKFLOW_COMPONENT_MISSING   | Error when a workflow component is missing a configuration. |
+| ERR_WORKFLOW_COMPONENT_NOT_FOUND | Error when a workflow component is not found.               |
+| ERR_WORKFLOW_EVENT_NAME_MISSING  | Error when the event name for a workflow is missing.        |
+| ERR_WORKFLOW_NAME_MISSING        | Error when the workflow name is missing.                    |
+| ERR_INSTANCE_ID_INVALID          | Error invalid workflow instance ID provided.                |
+| ERR_INSTANCE_ID_NOT_FOUND        | Error workflow instance ID not found.                       |
+| ERR_INSTANCE_ID_PROVIDED_MISSING | Error workflow instance ID was provided but missing.        |
+| ERR_INSTANCE_ID_TOO_LONG         | Error workflow instance ID exceeds allowable length.        |
+
+### State Management (Building Block)
```
Comment (msfussell) [inline daprdocs/content/en/reference/api/error_codes.md]:
```suggestion
### State Management API
```

File: daprdocs/content/en/reference/api/error_codes.md [hunk 6]
```diff
@@ -7,43 +7,150 @@ weight: 1400
...
+| ERR_STATE_DELETE                      | Error deleting a state from state store.                                  |
+| ERR_STATE_SAVE                        | Error saving a state in state store.                                      |
+| ERR_STATE_TRANSACTION                 | Error encountered during state transaction.                               |
+| ERR_STATE_BULK_GET                    | Error performing bulk retrieval of state entries.                         |
+| ERR_STATE_QUERY                       | Error querying the state store.                                           |
+| ERR_STATE_STORE_NOT_CONFIGURED        | Error state store is not configured.                                      |
+| ERR_STATE_STORE_NOT_SUPPORTED         | Error state store is not supported.                                       |
+| ERR_STATE_STORE_TOO_MANY_TRANSACTIONS | Error exceeded maximum allowable transactions.                            |
+
+### Configuration (Building Block)
```
Comment (msfussell) [inline daprdocs/content/en/reference/api/error_codes.md]:
```suggestion
### Configuration API
```

File: daprdocs/content/en/reference/api/error_codes.md [hunk 7]
```diff
@@ -7,43 +7,150 @@ weight: 1400
...
+
+| Error Code                             | Description                                  |
+| -------------------------------------- | -------------------------------------------- |
+| ERR_CONFIGURATION_GET                  | Error retrieving configuration.              |
+| ERR_CONFIGURATION_STORE_NOT_CONFIGURED | Error configuration store is not configured. |
+| ERR_CONFIGURATION_STORE_NOT_FOUND      | Error configuration store not found.         |
+| ERR_CONFIGURATION_SUBSCRIBE            | Error subscribing to a configuration.        |
+| ERR_CONFIGURATION_UNSUBSCRIBE          | Error unsubscribing from a configuration.    |
+
+### Crypto (Building Block)
```
Comment (msfussell) [inline daprdocs/content/en/reference/api/error_codes.md]:
```suggestion
### Crypto API
```

File: daprdocs/content/en/reference/api/error_codes.md [hunk 8]
```diff
@@ -7,43 +7,150 @@ weight: 1400
...
+
+| Error Code                          | Description                                |
+| ----------------------------------- | ------------------------------------------ |
+| ERR_CRYPTO                          | General crypto building block error.       |
+| ERR_CRYPTO_KEY                      | Error related to a crypto key.             |
+| ERR_CRYPTO_PROVIDER_NOT_FOUND       | Error specified crypto provider not found. |
+| ERR_CRYPTO_PROVIDERS_NOT_CONFIGURED | Error no crypto providers configured.      |
+
+### Secrets (Building Block)
```
Comment (msfussell) [inline daprdocs/content/en/reference/api/error_codes.md]:
```suggestion
### Secrets API
```

File: daprdocs/content/en/reference/api/error_codes.md [hunk 9]
```diff
@@ -7,43 +7,150 @@ weight: 1400
...
+
+| Error Code                       | Description                                          |
+| -------------------------------- | ---------------------------------------------------- |
+| ERR_SECRET_STORES_NOT_CONFIGURED | Error that no secret store is configured.            |
+| ERR_SECRET_STORE_NOT_FOUND       | Error that specified secret store is not found.      |
+| ERR_SECRET_GET                   | Error retrieving the specified secret.               |
+| ERR_PERMISSION_DENIED            | Error access denied due to insufficient permissions. |
+
+### Pub/Sub (Building Block)
```
Comment (msfussell) [inline daprdocs/content/en/reference/api/error_codes.md]:
```suggestion
### Pub/Sub API
```

File: daprdocs/content/en/reference/api/error_codes.md [hunk 10]
```diff
@@ -7,43 +7,150 @@ weight: 1400
...
+| ERR_PUBSUB_FORBIDDEN        | Error message forbidden by access controls.              |
+| ERR_PUBSUB_CLOUD_EVENTS_SER | Error serializing Pub/Sub event envelope.                |
+| ERR_PUBSUB_EMPTY            | Error empty Pub/Sub.                                     |
+| ERR_PUBSUB_NOT_CONFIGURED   | Error Pub/Sub component is not configured.               |
+| ERR_PUBSUB_REQUEST_METADATA | Error with metadata in Pub/Sub request.                  |
+| ERR_PUBSUB_EVENTS_SER       | Error serializing Pub/Sub events.                        |
+| ERR_PUBLISH_OUTBOX          | Error publishing message to the outbox.                  |
+| ERR_TOPIC_NAME_EMPTY        | Error topic name for Pub/Sub message is empty.           |
+
+### Conversation (Building Block)
```
Comment (msfussell) [inline daprdocs/content/en/reference/api/error_codes.md]:
```suggestion
### Conversation API
```

File: daprdocs/content/en/reference/api/error_codes.md [hunk 11]
```diff
@@ -7,43 +7,150 @@ weight: 1400
...
+| Error Code                      | Description                                     |
+| ------------------------------- | ----------------------------------------------- |
+| ERR_INVOKE_OUTPUT_BINDING       | Error invoking an output binding.               |
+| ERR_DIRECT_INVOKE               | Error in direct invocation.                     |
+| ERR_CONVERSATION_INVALID_PARMS  | Error invalid parameters for conversation.      |
+| ERR_CONVERSATION_INVOKE         | Error invoking the conversation.                |
+| ERR_CONVERSATION_MISSING_INPUTS | Error missing required inputs for conversation. |
+| ERR_CONVERSATION_NOT_FOUND      | Error conversation not found.                   |
+
+### Distributed Lock (Building Block)
```
Comment (msfussell) [inline daprdocs/content/en/reference/api/error_codes.md]:
```suggestion
### Distributed Lock API
```

File: daprdocs/content/en/operations/observability/metrics/metrics-overview.md [hunk 12]
```diff
@@ -70,6 +70,38 @@ spec:
...
+```yaml
+apiVersion: dapr.io/v1alpha1
+kind: Configuration
+metadata:
+  name: tracing
+  namespace: default
+spec:
+  metrics:
+    enabled: true
+    recordErrorCodes: true
```
Comment (msfussell) [inline daprdocs/content/en/operations/observability/metrics/metrics-overview.md:87]: You also need to;
1)  update this table in configuration setting https://docs.dapr.io/operations/configuration/configuration-overview/#metrics And in that table point to this article for the details. 
2) Update the configuration schema to add this under metrics section here https://docs.dapr.io/reference/resource-specs/configuration-schema/

**Comment by msfussell** [comment]
@jake-engelberg - I realize that we need to make error codes concept much more visible to developers by adding a topic under https://docs.dapr.io/developing-applications/building-blocks/ called "API Error codes" describing what they are and also how you can get metrics for them (This can point to the error codes list and the metrics configuration)

Also in this article we should mention that each SDK can handle and return the error code in the language specific way. For example this is Go https://docs.dapr.io/developing-applications/sdks/go/go-client/#error-handling but we also need to discuss with each of the SDK maintainers how to handle these. I suggest raising an issue into each of SDK repos for this to track this (if my understanding is correct)

@hhunter-ms can help as needed. Hannah - This reference article https://docs.dapr.io/reference/api/error_codes/ should be called "API Error codes" in the index file. I did not realize this was buried under Dapr API reference.

File: daprdocs/content/en/reference/api/error_codes.md [hunk 13]
```diff
@@ -7,43 +7,150 @@ weight: 1400
 ---
 
 For http calls made to Dapr runtime, when an error is encountered, an error json is returned in http response body. The json contains an error code and an descriptive error message, e.g.
```
Comment (alicejgibbons) [inline daprdocs/content/en/reference/api/error_codes.md:9]: @jake-engelberg this is HTTP and gRPC?

File: daprdocs/content/en/operations/observability/metrics/metrics-overview.md (hunk 12 above)
Comment (msfussell) [reply daprdocs/content/en/operations/observability/metrics/metrics-overview.md:87]: @hhunter-ms  - Will do this. Do Jake you can ignore this comment.

File: daprdocs/content/en/reference/api/error_codes.md (hunk 13 above)
Comment (jake-engelberg) [reply daprdocs/content/en/reference/api/error_codes.md:9]: Yes, it is both and we'll update the wording soon to reflect the auto generated chart with both codes which at the moment sometimes differ:

https://github.com/dapr/dapr/pull/8346

**Comment by marcduiker** [comment]
@holopin-bot @jake-engelberg Thanks Jake!

**Pull request merged**